### ISSUE TITLE
Possibly incorrect behaviour in quasiquote

### ISSUE THREAD

**Issue by rxg** [racket/racket#3028]
I expected "(quasiquote (quasiquote unquote)" to produce "'`unquote"  but instead it errors.

For (imperfect) reference, Chez Scheme Version 9.5's behaviour is the analogue to the above (though its printer omits the leading "'")

Transcript form Dr. Racket below:

> (quasiquote unquote)
'unquote
> (quasiquote (quasiquote unquote))
unquote: expects exactly one expression in: (unquote)
>

**Comment by jbclements** [comment]
I think I agree with you that this seems a bit inconsistent. I think the point is clearer without the inner quasiquote, though. Specifically, 

```
(quasiquote (frog bunny unquote))
```

produces the same error, and seems inconsistent with 

```
(quasiquote unquote)
```

... in the same way, right?

**Comment by rxg** [comment]
Yes it looks like this is the same issue.  unquote  seems to behave oddly under quasiquote if it is not intended to be used in (unquote e) form

**Comment by rxg** [comment]
I've noticed another curious behaviour, which might be intended, but I think has significant implications:  what happens when an unquote-splicing appears as the cdr of a cons:

> (quasiquote (x y . (unquote-splicing (list 'z 'a))))
. unquote-splicing: invalid context within quasiquote in: (unquote-splicing (list (quote z) (quote a)))
> (quasiquote (quasiquote (x y . (unquote-splicing (list 'z 'a)))))
. unquote-splicing: invalid context within quasiquote in: (unquote-splicing (list (quote z) (quote a)))

The first above seems ... like a plausible design, but the second seems less than desirable.
For comparison, here is Chez Scheme's behaviour:
> (quasiquote (x y . (unquote-splicing (list 'z 'a))))
(x y unquote-splicing (list 'z 'a))
> (quasiquote (quasiquote (x y . (unquote-splicing (list 'z 'a)))))
`(x y unquote-splicing (list 'z 'a))
>

At first, I was a bit taken-aback by the behaviour in the first case, but I came to view its semantics as being that unquote-splicing is only special when it appears in a "special" position.  To my relief, the second case is consistent with the first.  Mind you I imagine that in some cases where an error would be helpful instead of embedding 'unquote-splicing as a symbol in the list, but being able to stick the symbol 'unquote-splicing in an s-expression while using quasiquote has its benefits (e.g. using quasiquote while implementing a quasiquote compiler...which is where I ran into these behaviours).

**Comment by soegaard** [comment]
@rxg  FWIW here is little note on quasiquotation. Thought you'd find it interesting:
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.309.227&rep=rep1&type=pdf

**Comment by rxg** [comment]
Thanks Jens!  Coincidentally enough, I literally just downloaded this paper this morning, and am planning to share it with my students (the context is that I'm teaching compilers, and am helping students understand that ANF translation pops up in a lot of places)!  Good to know I'm on the right path :).

**Comment by PetSerAl** [comment]
```racket
(quasiquote (frog bunny unquote))
```
is semantically equivalent to
```racket
(quasiquote (frog bunny . (unquote)))
```
which is incorrect usage of `unquote`. If you write it like this:
```racket
(quasiquote (frog bunny unquote (* 6 7)))
```
or
```racket
(quasiquote (frog bunny . unquote))
```
then Racket will accept it. Not sure if last thing is by design.

>but being able to stick the symbol 'unquote-splicing in an s-expression while using quasiquote has its benefits

Nothing stop you doing so:
```racket
(quasiquote (x y . ((unquote (quote unquote-splicing)) (list 'z 'a))))
```

**Comment by jbclements** [comment]
okay to close this issue? Okay, actually, I'm going to close it; feel free to re-open if not resolved.

**Comment by mflatt** [comment]
I'm not convinced that the current behavior is optimal (even considering compatibility), so I'll reopen.

**Comment by rxg** [comment]
> ```racket
> (quasiquote (frog bunny unquote))
> ```
> 
> is semantically equivalent to
> 
> ```racket
> (quasiquote (frog bunny . (unquote)))
> ```
> 
> which is incorrect usage of `unquote`.

I agree that it's *syntactically* equivalent, but I'm not sure what semantics is most desirable for it.  Racket and Chez disagree in practice, but what principles might underly those behaviours?  

 If you write it like this:
> 
> ```racket
> (quasiquote (frog bunny unquote (* 6 7)))
> ```
> 
> or
> 
> ```racket
> (quasiquote (frog bunny . unquote))
> ```
> 
> then Racket will accept it. Not sure if last thing is by design.
> 
> > but being able to stick the symbol 'unquote-splicing in an s-expression while using quasiquote has its benefits
> 
I agree.

**Comment by PetSerAl** [comment]
>I agree that it's *syntactically* equivalent

Racket syntax model do allow distinguish this two lines, so them not equivalent syntactically. Not for Racket at least.

**Comment by rxg** [comment]
> > I agree that it's _syntactically_ equivalent
> 
> Racket syntax model do allow distinguish this two lines, so them not equivalent syntactically. Not for Racket at least.

Ahh, my mistake.  And good to know, thanks!

**Comment by Xiphoseer** [comment]
> ```racket
> (quasiquote (frog bunny unquote))
> ```
> 
> is semantically equivalent to
> 
> ```racket
> (quasiquote (frog bunny . (unquote)))
> ```

I was quite confused by this at first and I think I know now why that is. Those are syntactically equivalent in the reader so they are identical (minus some source information) to the expander, so they have the same semantics.

```racket
> (read-syntax)(quasiquote (a . (unquote 'b)))
#<syntax:readline-input:19:13 (quasiquote (a unquote (quote b)))>
```

But the implied consequence is misleading. An `unquote` within a quasiquote is not actually expanded as a syntactic form (macro) so it doesn't need to abide by those rules. The definition of unquote is [a macro that always fails](https://github.com/racket/racket/blob/6380df8acacca0067152114a9ac866583b4a6f4b/racket/src/expander/expand/expr.rkt#L756-L759).

If it were a macro, it could not behave the way it does. Consider the following example, a macro that unconditionally produces the datum `'foo`

```racket
> (define-syntax x (lambda (x) #'0))
> (x 1)
0
> (number? . (x 1))
; number?: arity mismatch;
;  the expected number of arguments does not match the given number
;   expected: 1
;   given: 2
; [,bt for context]
```
Why is that? Because `(number? . (x 1))` is read as `(number? x 1)` which expands to `(number? 0 1)`. The expansion of `x` does not change based on the parts that follow it. That would be true for `unquote` too, if it wouldn't just fail at expansion anyways.

So how is unquote handled? Consider the following error message:

```racket
> (quasiquote (unquote))
; readline-input:13:12: unquote: expects exactly one expression
;   at: (unquote)
;   in: (quasiquote (unquote))
; [,bt for context]
```

The `unquote: expects exactly one expression` suggests that the `unquote` syntactic form is the source of the error. That not true: The stacktrace indicates that the error is raised in the implementation of [`quasiquote`]. The binding of `unquote` is only [used to check][check-unquote] whether a symbol is bound to the real `unquote`.

So then the question is: What are the semantics of `unquote` within quasiquote? Consider the following:
```racket
> (quasiquote unquote)
'unquote
> (quasiquote unquote 1)
; readline-input:52:0: quasiquote: bad syntax
;   in: (quasiquote unquote 1)
; [,bt for context]
> (quasiquote (unquote 1))
1
> (quasiquote (1 unquote 2))
'(1 . 2)
> (quasiquote (1 unquote . 2))
'(1 unquote . 2)
```
There are a few things to note here:
1. `quasiquote` fails early with `bad syntax` if there isn't exactly one parameter
2. The code tells us that an `unquote` is only recognized whenever it appears as the first element of a  [`stx-list?`]. 
3. This means that pairs that don't terminate with `null` or a syntax object don't recognize `unquote` at all. That explains why a simple `(quasiquote unquote)` produces a symbol.
4. When the term is a `stx-list?`, then an `unquote` must be followed by exactly one element, otherwise it fails.

```racket
> (quasiquote (unquote unquote unquote . 1))
'(unquote unquote unquote . 1)
> (quasiquote (unquote 1 1))
; readline-input:24:12: unquote: expects exactly one expression
;   at: (unquote 1 1)
;   in: (quasiquote (unquote 1 1))
; [,bt for context]
```

A more accurate representation of how an inline `unquote` relates to the dot-notation is that it expands to a dot-notation, not the other way around, as seen here:
```
> (quasiquote (1 unquote 2))
'(1 . 2)
> (quasiquote (1 unquote (list 2 3)))
'(1 2 3)
> '(1 . (2 3))
'(1 2 3)
```

There are also interesting interactions with nesting `quasiquote`. When a nested quasiquote is encountered the level is increased but there is no check whether it only has a single part following it. That is not the case for `unquote` which still enforces that exactly one part must follow and which doesn't have any effect when it doesn't appear in a syntax list.
```racket
> (quasiquote (quasiquote unquote 1))
'(quasiquote unquote 1)
> (quasiquote (quasiquote unquote unquote 1))
; readline-input:83:0: unquote: expects exactly one expression
;   at: (unquote unquote 1)
;   in: (quasiquote (quasiquote unquote unquote 1))
; [,bt for context]
> (quasiquote (quasiquote unquote (unquote 1)))
'(quasiquote unquote 1)
> (quasiquote (quasiquote unquote (2 unquote 1)))
'(quasiquote unquote (2 . 1))
> (quasiquote (quasiquote (unquote 1)))
'`,1
> (quasiquote (quasiquote (2 unquote 1)))
'`(2 unquote 1)
```
There also seems to be some inconsitencies between the expansion, printer or macro-stepper concerning these undefined uses of `unquote`:
```racket
> (quasiquote (quasiquote unquote 1))
;; racket
'(quasiquote unquote 1)
;; raco expand
'(quasiquote . ,1)
;; macro stepper
(#%app list (quote quasiquote) (quote unquote) (quote 1))
> (quasiquote (quasiquote unquote (unquote 1)))
;; racket
'(quasiquote unquote 1)
;; raco expand
'(quasiquote . ,,1)
;; macro stepper
(quote (quasiquote unquote 1))
```

Especially the second one seems to contain a bug as `(quasiquote (quasiquote unquote (unquote 1)))` and `'(quasiquote . ,,1)` don't produce the same value. This was tested with Racket 7.2 as that is what I have on my machine but then again the implementation of `quasiquote` hasn't changed since at least 2015.

### Final thoughts

Most of these issues would never appear in code that uses the syntactic sugar ``` ` ```, `,@` and `,` as those read as `(unquote X)` syntax lists. So I feel that changing or at least specifying the current behaviour would not be too much of a backwards-compatibility issue.

I would imagine that better error messages that communicate how `unquote` should be used would be of immediate benefit. It also seems a little weird that you can just turn off unquote by adding a `. foo` at the end of the list.

Also consider the following:
```racket
> (quasiquote (1 . ,(list 2 3)))
'(1 2 3)
> (quasiquote (1 ,@(list 2 3)))
'(1 2 3)
```

An `unquote` in a list already behaves somewhat like an `unquote-slicing` in parenthesis. It may be useful to change the current behavior to allow any `(... unquote X ...)` to behave as if it were `(... ,@X ...)` and provide a custom error message when there is no part after `unquote`.

[`stx-list?`]: https://docs.racket-lang.org/syntax/syntax-helpers.html?q=stx-cons#(def._((lib._syntax%2Fstx..rkt)._stx-list~3f))
[`quasiquote`]: https://github.com/racket/racket/blob/6b1dad21ebd103960822e6de7d45e9a85303982b/racket/collects/racket/private/qq-and-or.rkt#L274
[check-unquote]: https://github.com/racket/racket/blob/6b1dad21ebd103960822e6de7d45e9a85303982b/racket/collects/racket/private/qq-and-or.rkt#L265
